File: common-core/src/main/java/io/quartic/common/server/ResourceManagingEndpoint.java
package io.quartic.common.server;

import org.slf4j.Logger;

import javax.websocket.CloseReason;
import javax.websocket.Endpoint;
import javax.websocket.EndpointConfig;
import javax.websocket.Session;

import static org.slf4j.LoggerFactory.getLogger;

public abstract class ResourceManagingEndpoint<T> extends Endpoint {
    private static final Logger LOG = getLogger(ResourceManagingEndpoint.class);
    private static final String RESOURCE = "resource";

    @Override
    public final void onOpen(Session session, EndpointConfig config) {
        LOG.info("[{}] Open", session.getId());
        session.getUserProperties().put(RESOURCE, createResourceFor(session));
    }

    @SuppressWarnings("unchecked")
    @Override
    public final void onClose(Session session, CloseReason closeReason) {
        LOG.info("[{}] Close", session.getId());
        releaseResource((T) session.getUserProperties().get(RESOURCE));
    }

    protected abstract T createResourceFor(Session session);
    protected abstract void releaseResource(T resource);
}

File: common-core/src/main/kotlin/io/quartic/common/websocket/ResourceManagingEndpoint.kt
package io.quartic.common.websocket

import io.quartic.common.logging.logger
import javax.websocket.CloseReason
import javax.websocket.Endpoint
import javax.websocket.EndpointConfig
import javax.websocket.Session

// TODO: use delegation rather than abstract class
abstract class ResourceManagingEndpoint<T> : Endpoint() {
    private val LOG by logger()

    final override fun onOpen(session: Session, config: EndpointConfig) {
        LOG.info("[{}] Open", session.id)
        session.userProperties.put(RESOURCE, createResourceFor(session))
    }

    @Suppress("UNCHECKED_CAST")
    final override fun onClose(session: Session?, closeReason: CloseReason?) {
        LOG.info("[{}] Close", session!!.id)
        releaseResource(session.userProperties[RESOURCE] as T)
    }

    protected abstract fun createResourceFor(session: Session): T
    protected abstract fun releaseResource(resource: T)

    companion object {
        private val RESOURCE = "resource"
    }
}